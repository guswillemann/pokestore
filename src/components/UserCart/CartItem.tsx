import styled from 'styled-components';
import useUserCart, { ProductType } from '../../contexts/UserCartContext';
import { getThemeColor } from '../../theme/utils';

const CartItemWrapper = styled.li`
  @keyframes slideLeft {
    0% { transform: translateX(100%) }
    100% { transform: translateX(0) }
  }

  animation: slideLeft 500ms ease-in;

  display: grid;
  gap: 1rem;
  grid-template-columns: 4rem 1fr 10rem 1rem;
  height: 5rem;
  background-color: ${getThemeColor('background')};
  padding: 0.5rem;
  border-radius: ${({ theme }) => theme.borderRadius};

  p {
    align-self: center;
  }

  .price-tag {
    margin-left: auto;
  }

  button {
    background-color: transparent;
    border: none;
    cursor: pointer;
    font-size: 0;
  }

  button:hover, button:focus {
    opacity: 0.5;
  }
`;

type CartItemPropsTypes = {
  product: ProductType;
  index: number;
}

export default function CartItem({ product, index }: CartItemPropsTypes) {
  const { removeItem } = useUserCart();

  return (
    <CartItemWrapper>
      <img src={product.img} alt="Imagem do produto" />
      <p>{product.name}</p>
      <p className="price-tag">{`R$ ${product.price}`}</p>
      <button type="button" onClick={() => removeItem(index)}>
        <img src="/images/x-icon.svg" alt="X" aria-label="Remover item do carrinho" />
      </button>
    </CartItemWrapper>
  );
}
