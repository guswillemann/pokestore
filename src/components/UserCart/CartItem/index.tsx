import useUserCart, { ProductType } from '../../../contexts/UserCartContext';
import Button from '../../Button';
import ProductImage from '../../ProductImage';
import CartItemWrapper from './styles';

type CartItemPropsTypes = {
  product: ProductType;
  index: number;
}

export default function CartItem({ product, index }: CartItemPropsTypes) {
  const { removeItem } = useUserCart();

  return (
    <CartItemWrapper data-testid="cart-item">
      <ProductImage src={product.img} />
      <div className="item-info">
        <p className="item-name">{product.name}</p>
        <p className="item-price">{`$${product.price}`}</p>
        <Button
          type="button"
          onClick={() => removeItem(index)}
          aria-label="Remove item from cart"
        >
          remove
        </Button>
      </div>
    </CartItemWrapper>
  );
}
