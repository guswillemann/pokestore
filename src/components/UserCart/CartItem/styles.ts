import styled, { keyframes } from 'styled-components';
import { getThemeColor } from '../../../theme/utils';

const itemGrow = keyframes`
  from {
    transform: scale(0.75);
    opacity: 0;
  }
  to {
    transform: scale(1);
    opacity: 1;
  }
`;

const CartItemWrapper = styled.li`
  animation: ${itemGrow} 300ms ease-in-out forwards;

  display: flex;
  background-color: ${getThemeColor('background')};
  border-radius: ${({ theme }) => theme.borderRadius};
  padding: 1rem;

  & > img {
    max-width: 30%;
  }

  .item-info {
    display: grid;
    grid-template-areas: "name name" "price btn";
    grid-template-rows: 1fr 1fr;
    align-items: center;
    flex: 1;
  }

  .item-name {
    grid-area: name;
    text-transform: capitalize;
    font-weight: 700;
    font-size: 2rem;
  }

  .item-price {
    grid-area: price;
  }

  button {
    grid-area: btn;
  }
`;

export default CartItemWrapper;
