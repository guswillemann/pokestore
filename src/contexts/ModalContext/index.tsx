import { createContext, ReactNode, useContext, useState } from 'react';
import LockScroll from '../../components/LockScroll';
import Modal from './Modal';

type ModalContextTypes = {
  modalContents: ReactNode;
  activeModal: (contents: ReactNode) => void;
  toggleModal: () => void;
}

export const ModalContext = createContext({} as ModalContextTypes);

type ModalProviderProps = {
  children: ReactNode;
}

export function ModalProvider({ children }: ModalProviderProps) {
  const [modalContents, setModalContents] = useState<ReactNode>(null);
  const [isModalVisible, setIsModalVisible] = useState(false);

  const hasModalContents = Boolean(modalContents);

  function activeModal(contents: ReactNode) {
    setModalContents(contents);
    setIsModalVisible(true);
  }
  
  function toggleModal() {
    setIsModalVisible(!isModalVisible);
  }

  return (
    <ModalContext.Provider value={{
      modalContents,
      activeModal,
      toggleModal,
    }}>
      {isModalVisible && <LockScroll />}
      {hasModalContents && <Modal isVisible={isModalVisible} />}
      {children}
    </ModalContext.Provider>
  );
}

export default function useModal() {
  return useContext(ModalContext);
};
