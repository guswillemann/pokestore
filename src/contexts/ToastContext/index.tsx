import { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import Toast, { ToastObjectType } from './Toast';

type ToastContextTypes = {
  activeToast: (toastObj: ToastObjectType) => void;
}

const ToastContext = createContext({} as ToastContextTypes);

type ToastProviderProps = {
  children: ReactNode;
}

export function ToastProvider({ children }: ToastProviderProps) {
  const [toastObject, setToastObject] = useState<ToastObjectType | null>();
  
  function activeToast(newToastObject: ToastObjectType) {
    setToastObject(newToastObject);
  }
  
  useEffect(() => {
    if (!toastObject) return;
    const timeoutId = setTimeout(() => {setToastObject(null)}, 3000);
    return () => clearTimeout(timeoutId);
  }, [toastObject])
  
  return (
    <ToastContext.Provider value={{
      activeToast,
    }}>
      {toastObject && <Toast toastObject={toastObject} />}
      {children}
    </ToastContext.Provider>
  );
}

export default function useToast() {
  return useContext(ToastContext);
};
