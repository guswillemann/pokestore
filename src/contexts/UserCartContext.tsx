import { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import nookies from 'nookies';
import { useRouter } from 'next/router';
import useToast from './ToastContext';

export type ProductType = {
  id: string;
  name: string;
  price: string;
  img: string;
}

type UserCartContextTypes = {
  cartList: Array<ProductType>;
  addToCart: (product: ProductType) => void;
  resetCart: () => void;
  toggleCart: () => void;
  removeItem: (index: number) => void;
  isCartVisible: boolean;
}

export const UserCartContext = createContext({} as UserCartContextTypes);

type UserCartProviderProps = {
  children: ReactNode;
  cartListCookie: string;
  avaibleProducts: Array<ProductType>;
}

export const CART_LIST_COOKIE = 'CART_LIST_COOKIE';

export function UserCartProvider({ children, cartListCookie, avaibleProducts }: UserCartProviderProps) {
  const { pathname } = useRouter();
  const { activeToast } = useToast();
  const [isCartVisible, setIsCartVisible] = useState(false);
  const [cartList, setCartList] = useState<Array<ProductType>>([]);

  useEffect(() => {
      if (cartListCookie) {
        const cookie = cartListCookie || nookies.get()[CART_LIST_COOKIE];
        const cookieCartItems = cookie.split(',');
        const initialCartList = avaibleProducts.filter((product) => cookieCartItems.some((cartItem) => {
          return product.id === cartItem;
        }));
        setCartList(initialCartList);
      }
  }, [cartListCookie])

  function addToCart(product: ProductType) {
    const hasProduct = cartList.some((cartItem) => product.id === cartItem.id)
    if (!hasProduct) {
      const newCartList = [...cartList, product];
      setCartList(newCartList);
      
      const cartIdList = newCartList.reduce((list, item) => [...list, item.id], [] as string[]);
      nookies.set(null, CART_LIST_COOKIE, cartIdList.toString(), {
        path: pathname,
        maxAge: 60 * 60 * 24 * 7,
      })
      return
    }

    activeToast({
      message: 'Você já possui este item no carrinho',
      type: 'alert',
    })
  }

  function resetCart() {
    setCartList([]);
    nookies.destroy(null, CART_LIST_COOKIE, { path: pathname });
  }

  function toggleCart() {
    setIsCartVisible(!isCartVisible);
  }

  function removeItem(index: number) {
    const newCartList = [...cartList];
    newCartList.splice(index, 1);
    setCartList(newCartList);
  }

  return (
    <UserCartContext.Provider value={{
      cartList,
      addToCart,
      resetCart,
      toggleCart,
      removeItem,
      isCartVisible,
    }}>
      {children}
    </UserCartContext.Provider>
  );
}

export default function useUserCart() {
  return useContext(UserCartContext);
};
