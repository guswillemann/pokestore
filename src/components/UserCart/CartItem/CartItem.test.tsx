import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FC, ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';
import { testTheme } from '../../../../test/test-utils';
import { UserCartContext } from '../../../contexts/UserCartContext';
import CartItem from '../CartItem';

const testProduct = {
  id: 'id',
  name: 'testProduct',
  img: 'img',
  price: '20',
}

const removeItemFunction = jest.fn();

const Wrapper: FC = ({ children }) => (
  <ThemeProvider theme={testTheme}>
    <UserCartContext.Provider value={{
      removeItem: removeItemFunction,
    } as any}>
      {children}
    </UserCartContext.Provider>
  </ThemeProvider>
);

describe('<CartItem />', () => {
  let cartItem: ReactNode;
  beforeEach(() => {
    cartItem = render(
      <CartItem
        product={testProduct}
        index={1}
      />,
      { wrapper: Wrapper }
    );
  });

  it('should match the snapshot', () => {
    expect(cartItem).toMatchSnapshot();
  });

  describe('when the remove button is clicked', () => {
    it('should execute the remove item function', () => {
      const removeBtn = screen.getByLabelText('Remove item from cart');
      userEvent.click(removeBtn);
      expect(removeItemFunction).toBeCalled();
      expect(removeItemFunction).toBeCalledWith(1);
    });
  });
});
