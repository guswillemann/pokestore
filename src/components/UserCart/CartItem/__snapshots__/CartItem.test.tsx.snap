// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c1 {
  width: 100%;
  height: 2em;
  border: 2px solid darkred;
  border-radius: 8px;
  font-weight: 700;
  background-color: #FAFAFA;
  -webkit-transition: 300ms ease-in-out;
  transition: 300ms ease-in-out;
  -webkit-transition-property: background-color,color;
  transition-property: background-color,color;
}

.c1:not(:disabled):hover {
  background-color: darkred;
  color: white;
  cursor: pointer;
}

.c0 {
  -webkit-animation: evplCe 300ms ease-in-out forwards;
  animation: evplCe 300ms ease-in-out forwards;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #FAFAFA;
  border-radius: 8px;
  padding: 1rem;
}

.c0 > img {
  max-width: 30%;
}

.c0 .item-info {
  display: grid;
  grid-template-areas: "name name" "price btn";
  grid-template-rows: 1fr 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 .item-name {
  grid-area: name;
  text-transform: capitalize;
  font-weight: 700;
  font-size: 2rem;
}

.c0 .item-price {
  grid-area: price;
}

.c0 button {
  grid-area: btn;
}

<body>
    <div>
      <li
        class="c0"
        data-testid="cart-item"
      >
        <img
          alt="Imagem do produto"
          src="img"
        />
        <div
          class="item-info"
        >
          <p
            class="item-name"
          >
            testProduct
          </p>
          <p
            class="item-price"
          >
            $20
          </p>
          <button
            aria-label="Remove item from cart"
            class="c1"
            type="button"
          >
            remove
          </button>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="styles__CartItemWrapper-sc-2uxl39-0 cnWdjo"
      data-testid="cart-item"
    >
      <img
        alt="Imagem do produto"
        src="img"
      />
      <div
        class="item-info"
      >
        <p
          class="item-name"
        >
          testProduct
        </p>
        <p
          class="item-price"
        >
          $20
        </p>
        <button
          aria-label="Remove item from cart"
          class="styles__ButtonWrapper-sc-136ut5a-0 PmdNa"
          type="button"
        >
          remove
        </button>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
